packages/uploadthing/src/client.ts

```diff
@@ -136,58 +166,50 @@ const uploadFile = <
   },
   presigned: PresignedURLs[number],
 ) =>
-  Effect.gen(function* () {
-    const file = opts.files.find((f) => f.name === presigned.fileName);
-
-    if (!file) {
-      console.error("No file found for presigned URL", presigned);
-      return yield* new UploadThingError({
-        code: "NOT_FOUND",
-        message: "No file found for presigned URL",
-        cause: `Expected file with name ${presigned.fileName} but got '${opts.files.join(",")}'`,
-      });
-    }
-
-    opts.onUploadBegin?.({ file: file.name });
-    if ("urls" in presigned) {
-      yield* uploadMultipartWithProgress(file, presigned, opts);
-    } else {
-      yield* uploadPresignedPostWithProgress(file, presigned, opts);
-    }
-
-    const PollingResponse = S.Union(
-      S.Struct({
-        status: S.Literal("done"),
-        callbackData: S.Any as S.Schema<TServerOutput, any>,
-      }),
-      S.Struct({ status: S.Literal("still waiting") }),
-    );
-
-    let serverData: TServerOutput | null = null;
-    if (!opts.skipPolling) {
-      serverData = yield* fetchEffJson(presigned.pollingUrl, PollingResponse, {
+  Array.findFirst(opts.files, (_) => _.name === presigned.fileName).pipe(
+    Effect.tapError(() =>
+      Console.error("No file found for presigned URL", presigned),
+    ),
+    Effect.mapError(
+      () =>
+        new UploadThingError({
+          code: "NOT_FOUND",
+          message: "No file found for presigned URL",
+          cause: `Expected file with name ${presigned.fileName} but got '${opts.files.join(",")}'`,
+        }),
+    ),
+    Effect.tap((_) => opts.onUploadBegin?.({ file: _.name })),
+    Effect.tap((_) =>
+      "urls" in presigned
+        ? uploadMultipartWithProgress(_, presigned, opts)
+        : uploadPresignedPostWithProgress(_, presigned, opts),
+    ),
+    Effect.zip(
+      fetchEffUnknown(presigned.pollingUrl, {
         headers: { authorization: presigned.pollingJwt },
       }).pipe(
-        Effect.andThen((res) =>
-          res.status === "done"
-            ? Effect.succeed(res.callbackData)
-            : Effect.fail(new RetryError()),
+        Effect.filterOrDieMessage(
+          isPollingResponse,
+          "received a non PollingResponse from the polling endpoint",
         ),
+        Effect.filterOrFail(isPollDone, () => new RetryError()),
+        Effect.map((_) => _.callbackData),
         Effect.retry({
           while: (res) => res instanceof RetryError,
           schedule: exponentialBackoff,
         }),
-      );
-    }
-
-    return {
+        Effect.when(() => !opts.skipPolling),
+        Effect.map(Option.getOrNull),
+        Effect.map(unsafeCoerce<unknown, TServerOutput>),
```

Comment:
it's just an identity function `(_) => _`, but it's like a sore thumb in the code, telling you that you're making unsafe assumptions. Just a nice util to use over straight casting